src/Core/Services/Implementations/StripePaymentService.cs

```diff
@@ -1707,5 +1658,115 @@ private async Task VerifyAppleReceiptNotInUseAsync(string receiptOriginalTransac
                 }
             }
         }
+
+        private decimal GetBillingBalance(Stripe.Customer customer)
+        {
+            return customer != null ? customer.Balance / 100M : default;
+        }
+
+        private async Task<BillingInfo.BillingSource> GetBillingPaymentSourceAsync(Stripe.Customer customer)
+        {
+            if (customer == null)
+            {
+                return null;
+            }
+
+            if (customer.Metadata?.ContainsKey("appleReceipt") ?? false)
+            {
+                return new BillingInfo.BillingSource
+                {
+                    Type = PaymentMethodType.AppleInApp
+                };
+            }
+
+            if (customer.Metadata?.ContainsKey("btCustomerId") ?? false)
+            {
+                try
+                {
+                    var braintreeCustomer = await _btGateway.Customer.FindAsync(
+                        customer.Metadata["btCustomerId"]);
+                    if (braintreeCustomer?.DefaultPaymentMethod != null)
+                    {
+                        return new BillingInfo.BillingSource(
+                            braintreeCustomer.DefaultPaymentMethod);
+                    }
+                }
+                catch (Braintree.Exceptions.NotFoundException) { }
+            }
+
+            if (customer.InvoiceSettings?.DefaultPaymentMethod?.Type == "card")
+            {
+                return new BillingInfo.BillingSource(
+                    customer.InvoiceSettings.DefaultPaymentMethod);
+            }
+
+            if (customer.DefaultSource != null &&
+                (customer.DefaultSource is Stripe.Card || customer.DefaultSource is Stripe.BankAccount))
+            {
+                return new BillingInfo.BillingSource(customer.DefaultSource);
+            }
+
+            var paymentMethod = GetLatestCardPaymentMethod(customer.Id);
+            return paymentMethod != null ? new BillingInfo.BillingSource(paymentMethod) : null;
+        }
+
+        private async Task<Stripe.Customer> GetCustomerAsync(string gatewayCustomerId, bool expandForPayments = false)
+        {
+            if (string.IsNullOrWhiteSpace(gatewayCustomerId))
+            {
+                return null;
+            }
+
+            Stripe.Customer customer = null;
+            try
+            {
+                Stripe.CustomerGetOptions customerOptions = null;
+                if (expandForPayments)
+                {
+                    customerOptions = new Stripe.CustomerGetOptions();
+                    customerOptions.AddExpand("default_source");
+                    customerOptions.AddExpand("invoice_settings.default_payment_method");
```

Comment:
I wonder if we could extract these to a separate method, and have `Stripe.CustomerGetOptions` as a method argument instead. Would simplify it a fair bit.

```c#
var customer = GetCustomerAsync(id, InvoicePaymentMethod())
```

Reply:
Updated to simplify the method as suggested.